### DISCUSSION TITLE
Accessing remote IP in <script setup>

### DISCUSSION THREAD

**Discussion by abdullrahman02000** [nuxt/nuxt#17585] [Questions]
### Environment

Nuxt project info:                                                                                                                                          16:35:43

------------------------------
- Operating System: `Linux`
- Node Version:     `v17.0.0`
- Nuxt Version:     `3.0.0-27460489.53fbca7`
- Package Manager:  `npm@8.5.5`
- Builder:          `vite`
- User Config:      `serverMiddleware`
- Runtime Modules:  `-`
- Build Modules:    `-`
------------------------------

### Reproduction

// Only one file "app.vue" can reproduce the issue
```

<script setup lang="ts">
enum Gender {
  GIRL = 'Girl',
      BOY = 'Boy',
      UNISEX = 'Unisex'
}
const options = reactive({
  gender: 'Girl',
});

const app = useNuxtApp()
const ipAddress = app.ssrContext.req.socket.remoteAddress
</script>
<template>
<div>
  <button v-for="(genderType,idx) in Gender" :key="idx" @click="options.gender=genderType" :class="{active: genderType==options.gender}">{{genderType}}</button>
  <button>Not effected</button>
</div>
</template>

<style scoped>
  .active {
    background: rgb(255, 77,23)
  }
</style>

```

### Development Environment
When the code is executed in dev mode, it will cause that the 3 buttons "Girl", "Boy", "Unisex" being not interactive
If you try to comment the line:
<strong>const ipAddress = app.ssrContext.req.socket.remoteAddress</strong>
The website will work normally
There might be exception happening while trying to access ip Address <strong>"returns undefined in dev mode"</strong>
I tried accessing ip address in server route and the elements are not getting effected
Also, the fourth button of title "Not effected", may not get effected anyway
If you may try adding checkbox with v-model, It is not effected
But in my case, buttons with <strong>reactive</strong> or <strong>ref</strong> as I have also tried will + v-for on enum and change value inside <strong>ref</strong> or <strong>reactive</strong> when <strong>@click</strong> will hang up if you try accessing ip address from nuxt ssr context inside <script setup>

### Production Environment
In production mode, the four buttons will be rendered normally:
`<button class="active" data-v-6a8b2cd5>Girl</button><button class="" data-v-6a8b2cd5>Boy</button><button class="" data-v-6a8b2cd5>Unisex</button><!--]--><button data-v-6a8b2cd5>Not effected</button>`
<strong>But once the page is loaded, 3 buttons will be hidden and remains 1 buttons with class "active" visible</strong>
If you try removing button "Not Effected" then 3 buttons are normally rendered:
`<button class="active" data-v-a4bb4f60>Girl</button><button class="" data-v-a4bb4f60>Boy</button><button class="" data-v-a4bb4f60>Unisex</button>`
But they are all hidden before and after the page complete loading

**Comment by huang-julien** [comment]
https://v3.nuxtjs.org/guide/going-further/internals#the-nuxtapp-interface
That's because an error has been thrown in the setup function which should instantiate the component. ssrContext is available only on server side. Wrap your `ipAddress` with `if(process.server)`. About the prod issue, that's because the error is thrown only client-side since ssrContext is available on server, causing a hydration error then vue js tries to perform a full-render client side.

**Comment by abdullrahman02000** [comment]
Thank you so much for your comment, but, Why not to render the line in server side only? It is better than check because it should be logically rendered on server since it is not logical to read ip address in client side, like how many things in <script setup> is rendered on server only depending on which part is trying to access and how they are called.

**Comment by huang-julien** [comment]
because that's how it works (at least on nuxt), we write code and components that should work on both server and client side. It means that we should be aware that some data, helpers or methods are not available on browsers or nodeJS depending on which side the code is executed. if you debug in dev mode, you can see that on server side, `if(process.server)` is compiled into `if(true)`  and `if(process.client)` -> `if(false)`. if you debug in client side `if(process.server)` is compiled into `if(false)` and  `if(process.client)` -> `if(true)`
